File: src/test/arm.py
from asm.testsource import *  # pylint: disable=W0614

class ArmTestSource(TestSource):

    @property
    def name(self) -> str:
        return 'arm'

    @property
    def test_cases(self) -> TestCases:
        yield TestCase('should encode single cps instruction', [
            self.make_call('cps', 'Mode::USR')
        ], bytearray(b'\x10\x00\x02\xf1'))

File: languages/cpp/test/arm.cpp
#ifndef ASM_ALL_TESTS
  #define CATCH_CONFIG_MAIN
#endif

#include <sstream>
#include "catch.hpp"
#include "../src/arm.cpp"

using Catch::Matchers::Equals;

using namespace std::string_literals;

TEST_CASE("arm tests", "[arm]") {
    std::ostringstream buf;

    SECTION("should encode single cps instruction") {
        arm::cps(buf, arm::Mode::USR);

        REQUIRE( buf.tellp() == 4 );
        REQUIRE_THAT(buf.str(), Equals("\x10\x00\x02\xf1"s));
    }
}
